### PR TITLE
Chat - Saving information about the user who wrote a message

### LINKED ISSUE #1361: Chat messages not restored on refresh
`GET /remoteaccess/chat` throws a silent error like in #1355, so the messages aren't retrieved.

```
Traceback (most recent call last):
  File "/home/bocciare/tools/micromamba/envs/mxcubeweb/lib/python3.10/site-packages/flask/app.py", line 880, in full_dispatch_request rv = self.dispatch_request()
  File "/home/bocciare/tools/micromamba/envs/mxcubeweb/lib/python3.10/site-packages/flask/app.py", line 865, in dispatch_request return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args) # type: ignore[no-any-return]
  File "/home/bocciare/tools/micromamba/envs/mxcubeweb/lib/python3.10/site-packages/flask_security/decorators.py", line 432, in decorated_view return current_app.ensure_sync(fn)(*args, **kwargs)
  File "/home/bocciare/workspace/mxcubeweb/mxcubeweb/routes/ra.py", line 215, in get_all_mesages return jsonify({"messages": app.chat.get_all_messages()})
  File "/home/bocciare/workspace/mxcubeweb/mxcubeweb/core/components/chat.py", line 40, in get_all_messages "username": _m.username,

AttributeError: 'Message' object has no attribute 'username' 
```

### PR DESCRIPTION
**Pull request by marcus-oscarsson**
The database schema for our chat is a bit limited in the relation messages to users. There is currently not really a way to know who authored a message. From a database model perspective one should probably have a author column with the author id in the `message` table or some other design of the mapping message to users. However I'm not sure its needed for our quite simple chat. 

Currently missing in this PR is a field `read` to signify if the message was read by the receiver or not.

This PR Closes #1361.

### CODE REVIEW

File: mxcubeweb/core/models/usermodels.py [hunk 1]
```diff
@@ -43,6 +43,10 @@ class Message(Base):
     id = Column(Integer(), primary_key=True)
     at = Column(DateTime())
     message = Column(Text())
+    read = Column(Boolean(False))
```
Comment (marcus-oscarsson) [inline mxcubeweb/core/models/usermodels.py:46]: Added the following columns to keep track of the author and if the message was read

File: ui/src/components/Main.jsx [hunk 2]
```diff
@@ -75,6 +90,7 @@ class Main extends React.Component {
   }
 
   render() {
+    setBadgeCount(this.props.remoteAccess.chatMessageCount);
```
Comment (marcus-oscarsson) [inline ui/src/components/Main.jsx]: the chat widget (`react-chat-widget`) now apparently handles the message count internally (see setBadgeCount, https://github.com/Wolox/react-chat-widget) so we need to override that behavior

File: ui/src/reducers/remoteAccess.js [hunk 3]
```diff
@@ -36,8 +36,11 @@ function remoteAccessReducer(state = INITIAL_STATE, action = {}) {
     case 'RESET_CHAT_MESSAGE_COUNT': {
       return { ...state, chatMessageCount: 0 };
     }
-    case 'INC_CHAT_MESSAGE_COUNT': {
-      return { ...state, chatMessageCount: state.chatMessageCount + 1 };
+    case 'SET_CHAT_MESSAGE_COUNT': {
```
Comment (marcus-oscarsson) [inline ui/src/reducers/remoteAccess.js]: Simply made it possible to pass the number of unread messages

File: ui/src/components/Main.jsx [hunk 4]
```diff
@@ -48,13 +56,20 @@ class Main extends React.Component {
...
         } else {
           addResponseMessage(
             `${entry.date} **${entry.nickname}:** \n\n ${entry.message}`,
           );
+
+          if (!entry.read) {
+            unread++;
+          }
+
+          this.props.setChatMessageCount(unread);
```
Comment (axelboc) [inline ui/src/components/Main.jsx]: Probably should be called after the loop?

File: ui/src/reducers/remoteAccess.js (hunk 3 above)
Comment (axelboc) [reply ui/src/reducers/remoteAccess.js]: The naming doesn't make it clear that the message count is being incremented. I think the `INC` prefix was fine before; "to increment" doesn't necessarily mean by 1.

File: ui/src/components/Main.jsx (hunk 2 above)
Comment (axelboc) [reply ui/src/components/Main.jsx]: This should be called in `componentDidUpdate` (or `useEffect` if it was a function component).

If you like, I can open a sub-PR to refactor `Main` into a function component (I had already done so yesterday actually, but hadn't pushed it yet).

File: ui/src/api/remoteAccess.js [hunk 5]
```diff
@@ -51,3 +51,7 @@ export function fetchChatMessages() {
 export function sendChatMessage(message, username) {
   return endpoint.post({ message, username }, '/chat').res();
 }
+
+export function setAllMessagesRead(message, username) {
```
Comment (axelboc) [inline ui/src/api/remoteAccess.js]: `sendSetAllMessagesRead`

File: ui/src/components/Main.jsx (hunk 2 above)
Comment (marcus-oscarsson) [reply ui/src/components/Main.jsx]: Actually, that for some reason did not work. I had to move `setBadeCount` to render to make it work. Wonderful for the functional component, I was hesitating of doing it as well. Ill let you continue if you already started :)

Comment (marcus-oscarsson) [reply ui/src/components/Main.jsx]: Ill quickly address some of your comments below and then you can go for it

File: ui/src/api/remoteAccess.js (hunk 5 above)
Comment (marcus-oscarsson) [reply ui/src/api/remoteAccess.js]: Oh yes

File: ui/src/components/Main.jsx (hunk 4 above)
Comment (marcus-oscarsson) [reply ui/src/components/Main.jsx]: Yes, of course

File: ui/src/reducers/remoteAccess.js (hunk 3 above)
Comment (marcus-oscarsson) [reply ui/src/reducers/remoteAccess.js]: I see,
 
> "to increment" doesn't necessarily mean by 1.

True, I somehow found it nicer with SET but INC is just as fine

File: ui/src/components/Main.jsx (hunk 2 above)
Comment (marcus-oscarsson) [reply ui/src/components/Main.jsx]: Should be ok now :)

Comment (axelboc) [reply ui/src/components/Main.jsx]: Thanks, I'll take over.

**Pull request merged**